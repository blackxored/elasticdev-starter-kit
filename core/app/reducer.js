//@flow
import type {
  Action,
  AppState,
  AppStartedAction,
  AppOnlineAction,
  AppErrorAction,
  Deps,
} from '../types';
import { assoc, merge } from 'ramda';

const APP_ERROR = 'esk/app/ERROR';
const APP_ONLINE = 'esk/app/ONLINE';
const APP_STARTED = 'esk/app/STARTED';

export const initialState: AppState = {
  error: null,
  online: false,
  started: false,
};

const reducer = (state: AppState = initialState, action: Action): AppState => {
  // Map all app errors into state.app.error.
  // In React Native, we show errors in one nicely animated unobtrusive alert.
  // In the browser, we prefer local error messages rendering.
  // action.error is FSA's.
  if (action.error === true && action.payload) {
    state = assoc('error', action.payload, state);
  }

  switch (action.type) {
    case APP_ERROR: {
      return assoc('error', action.payload, state);
    }
    case APP_ONLINE: {
      return merge(state, { error: null, online: action.payload.online });
    }
    case APP_STARTED: {
      return assoc('started', true, state);
    }
    default:
      return state;
  }
};

export const appStarted = (): AppStartedAction => ({
  type: APP_STARTED,
});

export const appError = (error: Error): AppErrorAction => ({
  type: APP_ERROR,
  payload: error,
  error: true,
});

export const appOnline = (online: boolean): AppOnlineAction => ({
  type: APP_ONLINE,
  payload: { online },
});

const appStartedEpic = (action$: any, deps: Deps) => {
  // TODO
};

export const epics = [
  /* TODO: appStartedEpic */
];
export default reducer;
