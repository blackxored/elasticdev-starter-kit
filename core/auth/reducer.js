// @flow
import type {
  Action,
  AuthState,
  AuthProviderName,
  Deps,
  LoginRequestAction,
  LoginFailureAction,
  LoginSuccessAction,
  LogoutAction,
} from '../types';
import { merge } from 'ramda';
import { Observable } from 'rxjs';
import { resetNavigation } from '../navigation/actions';

const LOGIN_REQUEST = 'esk/auth/LOGIN_REQUEST';
const LOGIN_SUCCESS = 'esk/auth/LOGIN_SUCCESS';
const LOGIN_FAILURE = 'esk/auth/LOGIN_FAILURE';
const AUTH_TOKENS_CHANGED = 'esk/auth/TOKENS_CHANGED';
const LOGOUT = 'esk/auth/LOGOUT';

export const initialState = {
  isAuthenticated: false,
  authorizationToken: null,
  refreshToken: null,
  errorMessage: '',
  authProvider: null,
  isFetching: false,
};

const reducer = (
  state: AuthState = initialState,
  action: Action,
): AuthState => {
  switch (action.type) {
    case LOGIN_REQUEST: {
      return merge(state, {
        isFetching: true,
        authProvider: action.payload.provider,
      });
    }
    case LOGIN_SUCCESS: {
      return merge(state, {
        isAuthenticated: true,
        isFetching: false,
        authorizationToken: action.payload.authorizationToken,
        refreshToken: action.payload.refreshToken,
      });
    }
    case LOGIN_FAILURE: {
      return merge(state, {
        isAuthenticated: false,
        isFetching: false,
        errorMessage: action.payload.message,
        authorizationToken: null,
        refreshToken: null,
      });
    }
    case AUTH_TOKENS_CHANGED: {
      return merge(state, {
        authorizationToken: action.payload.authorizationToken,
        refreshToken: action.payload.refreshToken,
      });
    }
    case LOGOUT: {
      return initialState;
    }
    default: {
      return state;
    }
  }
};

export function loginRequest(provider: AuthProviderName, providerData: mixed = {}): LoginRequestAction {
  return {
    type: LOGIN_REQUEST,
    payload: {
      provider,
      providerData,
    },
  };
}

type LoginSuccessParams = {
  authorizationToken: string,
  refreshToken: string,
  authProvider: AuthProviderName,
}

export function loginSuccess({
  authorizationToken,
  refreshToken,
  authProvider,
}: LoginSuccessParams): LoginSuccessAction {
  return {
    type: LOGIN_SUCCESS,
    payload: {
      authorizationToken,
      refreshToken,
    },
    meta: {
      authProvider,
    },
  };
}

export function authTokensChanged(
  authorizationToken: ?string,
  refreshToken: ?string,
) {
  return {
    type: AUTH_TOKENS_CHANGED,
    payload: {
      authorizationToken,
      refreshToken,
    },
  };
}


export function loginFailure(err: Error): LoginFailureAction {
  return {
    type: LOGIN_FAILURE,
    payload: err,
    error: true,
  };
}

export function logout(): LogoutAction {
  return {
    type: LOGOUT,
  };
}

const logoutEpic = (action$: any) => action$
  .filter((action: Action) => action.type === LOGOUT)
  .mapTo(resetNavigation('login'));

export const epics = [logoutEpic];

export default reducer;
