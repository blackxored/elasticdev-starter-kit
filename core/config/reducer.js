// @flow
import type { ConfigState } from '../types';

const initialState = {
  appName: '',
  appVersion: '',
  apiUrl: '',
  sentryUrl: '',
};

const reducer = (state: ConfigState = initialState): ConfigState => state;

export default reducer;
