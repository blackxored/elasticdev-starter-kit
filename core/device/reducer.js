// @flow
import type { DeviceInfoResponseAction, DeviceState } from '../types';
import { merge } from 'ramda';

type Action = DeviceInfoResponseAction;

const DEVICE_INFO_RESPONSE = 'esk/device/INFO_RESPONSE';

const initialState: DeviceState = {
  host: '',
};

const reducer = (
  state: DeviceState = initialState,
  action: Action,
): DeviceState => {
  switch (action.type) {
    case DEVICE_INFO_RESPONSE: {
      return merge(state, action.payload);
    }
    default: {
      return state;
    }
  }
};

export const deviceInfoResponse = (
  payload: mixed,
): DeviceInfoResponseAction => ({
  type: DEVICE_INFO_RESPONSE,
  payload,
});

export default reducer;
