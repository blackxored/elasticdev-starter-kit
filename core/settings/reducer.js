// @flow
import type { SettingsSetValueAction, SettingsState } from '../types';
import { assocPath } from 'ramda';

type Action = SettingsSetValueAction;

const SET_VALUE = 'esk/settings/SET_VALUE';

export const initialState: SettingsState = {
  notifications: false,
};

const reducer = (state: SettingsState = initialState, action: Action) => {
  switch (action.type) {
    case SET_VALUE: {
      return assocPath(action.payload.path, action.payload.value, state);
    }
    default: {
      return state;
    }
  }
};

export const setSettingsValue = (path: Array<string>, value: any) => ({
  type: SET_VALUE,
  payload: { path, value },
});

export default reducer;
