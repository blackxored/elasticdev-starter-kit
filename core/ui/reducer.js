// @flow
import type { ToggleNetworkIndicatorAction, UIState } from '../types';
import { assoc } from 'ramda';

type Action = ToggleNetworkIndicatorAction;

const TOGGLE_NETWORK_ACTIVITY_INDICATOR = 'esk/ui/TOGGLE_NETWORK_ACTIVITY_INDICATOR';

export const initialState = {
  showNetworkIndicator: false,
};

export const toggleNetworkActivityIndicator = (shouldShow: boolean): ToggleNetworkIndicatorAction => ({
  type: TOGGLE_NETWORK_ACTIVITY_INDICATOR,
  payload: shouldShow,
});

const reducer = (state: UIState = initialState, action: Action): UIState => {
  switch (action.type) {
    case TOGGLE_NETWORK_ACTIVITY_INDICATOR: {
      return assoc('showNetworkIndicator', action.payload, state);
    }
    default: {
      return state;
    }
  }
};

export default reducer;
