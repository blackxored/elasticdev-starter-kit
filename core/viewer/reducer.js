// @flow
import type { Action, ViewerState } from '../types';
import { path } from 'ramda';

export const initialState = null;

const reducer = (state: ?ViewerState = initialState, action: Action) => {
  const user = path(['result', 'data', 'viewer', 'user'], action);

  switch (action.type) {
    case 'APOLLO_QUERY_RESULT': {
      return user || state;
    }
    case 'LOGOUT': {
      return null;
    }
    default: {
      return state;
    }
  }
};

export default reducer;
