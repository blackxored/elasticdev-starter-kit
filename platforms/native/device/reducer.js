// @flow
import type {
  Action,
  Deps,
  DeviceState,
  MobileDeviceState,
} from '@esk/core/types';
import { merge } from 'ramda';
import { Observable } from 'rxjs';
import DeviceInfo from 'react-native-device-info';
import { deviceInfoResponse } from '@esk/core/device/reducer';

const DEVICE_INFO_RESPONSE = 'esk/device/INFO_RESPONSE';

const initialState: MobileDeviceState = {
  host: null,
  isReactNative: true,
  deviceModel: null,
  deviceId: null,
  deviceOS: null,
  appVersion: null,
  systemVersion: null,
  bundleId: null,
  locale: null,
};

const reducer = (
  state: MobileDeviceState = initialState,
  action: Action,
): DeviceState => {
  switch (action.type) {
    case DEVICE_INFO_RESPONSE: {
      return merge(state, action.payload);
    }
    default: {
      return state;
    }
  }
};

const DAY = 86400000;

const deviceInfoEpic = (action$: any, deps: Deps) =>
  action$.ofType('esk/app/STARTED').switchMap(() => {
    const lastTimestamp = deps.getState().device.timestamp;

    if (!lastTimestamp || deps.now() - lastTimestamp > DAY) {
      return Observable.of(
        deviceInfoResponse({
          timestamp: deps.now(),
          host: DeviceInfo.getUniqueId(),
          deviceModel: DeviceInfo.getModel(),
          deviceId: DeviceInfo.getDeviceId(),
          deviceOS: DeviceInfo.getSystemName(),
          appVersion: DeviceInfo.getReadableVersion(),
          systemVersion: DeviceInfo.getSystemVersion(),
          bundleId: DeviceInfo.getBundleId(),
          locale: DeviceInfo.getDeviceLocale(),
        }),
      );
    }

    return Observable.of();
  });

export const epics = [deviceInfoEpic];

export default reducer;
