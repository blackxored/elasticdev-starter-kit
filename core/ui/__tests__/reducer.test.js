import reducer, {
  initialState,
  toggleNetworkActivityIndicator,
} from '../reducer';

describe('ui reducer', () => {
  it('has an initial state', () => {
    expect(reducer(undefined, { type: '@@INIT' })).toMatchSnapshot();
  });

  it('handles TOGGLE_NETWORK_ACTIVITY_INDICATOR', () => {
    const showingState = reducer(
      initialState,
      toggleNetworkActivityIndicator(true),
    );
    expect(showingState).toMatchSnapshot();
    expect(
      reducer(showingState, toggleNetworkActivityIndicator(false)),
    ).toMatchSnapshot();
  });
});
