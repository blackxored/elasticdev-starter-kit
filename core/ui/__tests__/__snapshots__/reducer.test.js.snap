// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui reducer handles TOGGLE_NETWORK_ACTIVITY_INDICATOR 1`] = `
Object {
  "showNetworkIndicator": true,
}
`;

exports[`ui reducer handles TOGGLE_NETWORK_ACTIVITY_INDICATOR 2`] = `
Object {
  "showNetworkIndicator": false,
}
`;

exports[`ui reducer has an initial state 1`] = `
Object {
  "showNetworkIndicator": false,
}
`;
